Cap snakemake cores by available CPUs duing tests

peterjc opened the pull request:
This is important on older machines and the GitHub Actions works (which typically have only 2 to 4 CPUs available), and when running under an HPC cluster system like SLURM.

This partly addresses point one of #62.

peterjc commented:
The high level GitHub action timings seem little changed, perhaps even a fraction slower?

I was hoping this might be fractionally faster with less over-allocation of resources like memory and handles due to running less copies of ``nucmer`` etc at once.

peterjc commented:
As expected on our 24 core Linux server, no meaningful change (~1 min with workers, ~4 mins single-threaded aside the command line tools invoked):

```
$ time pytest -v -n auto
===================================== test session starts ======================================
platform linux -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0 -- /home/pjacock/miniforge3/envs/pyani-plus_py312/bin/python3.12
cachedir: .pytest_cache
rootdir: /home/pjacock/repositories/pyani-plus
configfile: pyproject.toml
plugins: cov-5.0.0, xdist-3.6.1
24 workers [7 items]      
scheduling tests via LoadScheduling

tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_filter 
tests/snakemake/test_snakemake_fastani_workflow.py::test_snakemake_rule_fastani 
tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_delta 
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_coords 
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_diff 
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_delta 
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_filter 
[gw6] [ 14%] PASSED tests/snakemake/test_snakemake_fastani_workflow.py::test_snakemake_rule_fastani 
[gw0] [ 28%] PASSED tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_filter 
[gw1] [ 42%] PASSED tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_delta 
[gw2] [ 57%] PASSED tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_delta 
[gw3] [ 71%] PASSED tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_filter 
[gw5] [ 85%] PASSED tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_coords 
[gw4] [100%] PASSED tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_diff 

======================================= warnings summary =======================================
tests/snakemake/test_snakemake_fastani_workflow.py::test_snakemake_rule_fastani
tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_filter
tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_delta
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_delta
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_filter
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_coords
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_diff
  /home/pjacock/miniforge3/envs/pyani-plus_py312/lib/python3.12/site-packages/stopit/__init__.py:10: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== 7 passed, 7 warnings in 61.44s (0:01:01) ===========================

real	1m1.720s
user	32m18.497s
sys	0m40.623s
```

```
$ time pytest -v 
===================================== test session starts ======================================
platform linux -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0 -- /home/pjacock/miniforge3/envs/pyani-plus_py312/bin/python3.12
cachedir: .pytest_cache
rootdir: /home/pjacock/repositories/pyani-plus
configfile: pyproject.toml
plugins: cov-5.0.0, xdist-3.6.1
collected 7 items                                                                              

tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_filter PASSED       [ 14%]
tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_delta PASSED        [ 28%]
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_delta PASSED     [ 42%]
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_filter PASSED    [ 57%]
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_diff PASSED [ 71%]
tests/snakemake/test_snakemake_dnadiff_workflow.py::test_snakemake_rule_show_coords PASSED [ 85%]
tests/snakemake/test_snakemake_fastani_workflow.py::test_snakemake_rule_fastani PASSED   [100%]

======================================= warnings summary =======================================
tests/snakemake/test_snakemake_anim_workflow.py::test_snakemake_rule_filter
  /home/pjacock/miniforge3/envs/pyani-plus_py312/lib/python3.12/site-packages/stopit/__init__.py:10: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== 7 passed, 1 warning in 230.40s (0:03:50) ===========================

real	3m50.756s
user	26m6.106s
sys	0m29.169s
```

widdowquinn commented:
> The high level GitHub action timings seem little changed, perhaps even a fraction slower?
> 
> I was hoping this might be fractionally faster with less over-allocation of resources like memory and handles due to running less copies of `nucmer` etc at once.

It could have gone either way, and I wouldn't have been surprised. Faster could have been due to less overallocation, as you say. Slower could be because the trade-off in overallocation was more efficient in overall use of CPU.

Ultimately, with this approach we're trying to avoid throwing that weird `OSError` when there's actually nothing wrong with the code, so I'm OK with a little bit slower, so long as we don't get an intermittent, hard to diagnose error again.

peterjc commented:
Agreed - avoiding intermittent errors is more important that marginal speedups.

peterjc commented:
The quad-core iMac seems to have hyperthreading enabled, it appears to have 8 cores in Activity Monitor etc: 

```
$ python -c "import os; print(os.cpu_count())"
8
```

i.e. This PR should have no effect on this machine.

widdowquinn commented on tests/snakemake/test_snakemake_anim_workflow.py:
```diff
@@ -64,7 +65,7 @@ def config_filter_args(
         "outdir": anim_nucmer_targets_filter_outdir,
         "indir": str(input_genomes_small),
-        "cores": 8,
+        "cores": min(8, available_cores),
```
Given the number of times `min(8, available_cores)` comes up, along with some other configurable settings that apply across multiple tests, can we assign this in something like `tools.py`/`config.py` under the `tests` directory, and import it instead?

peterjc replied:
Or as yet another pytest fixture?

Rather than in the tests, I'd lean to adding a helper function within pyani-plus itself. Something like https://github.com/peterjc/thapbi-pict/blob/v1.0.13/thapbi_pict/__main__.py#L42 although probably under a new ``tools.py`` or ``utils.py`` module?

widdowquinn replied:
I think you're right - a package-level module like `utils.py` would be a more sensible home. We could use that as a default for the CLI.

peterjc replied:
For now I replaced the ``available_cores`` fixture with a ``snakemake_cores`` fixture.

peterjc commented:
Rebased with simplified fixture.

Having read https://docs.pytest.org/en/7.1.x/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files I introduced ``tests/snakemake/conftest.py`` initially with just the new fixture, but potentially most/all of the existing fixtures are snakemake specific and should be there?

widdowquinn commented on tests/snakemake/conftest.py:
```diff
@@ -41,7 +41,7 @@
 
 
-@pytest.fixture()
+@pytest.fixture
```
Either `pytest.fixture` or `pytest.fixture()` work - I thought the convention was to have the parentheses present even if no arguments are passed, and we have that elsewhere. I'd prefer consistency if possible, so we should decide on one or other option.

Is there an argument for preferring one over the other, in general?

peterjc replied:
Something in ruff apparently thinks without is better - I did not look into the rational, but they will be consistent.

peterjc replied:
See 80fa7a908debc5c8605028c34ea00d30250afca2 from #66 which changed the rest

widdowquinn replied:
So it did… as [ruff notes](https://docs.astral.sh/ruff/rules/pytest-fixture-incorrect-parentheses-style/), what matters is consistency and it's a style choice.

widdowquinn commented:
That's great - thanks @peterjc !
